e2e/node.go

```diff
@@ -0,0 +1,53 @@
+package e2e_test
+
+import (
+	"context"
+	"fmt"
+
+	nodev1 "k8s.io/api/node/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/yaml"
+)
+
+const (
+	wasmSpinRuntime   = "spin"
+	wasmSlightRuntime = "slight"
+)
+
+func ensureWasmRuntimeClasses(ctx context.Context, kube *kubeclient) error {
+	className := fmt.Sprintf("wasmtime-%s", wasmSpinRuntime)
+
+	var runtimeClassList nodev1.RuntimeClassList
+	if err := kube.dynamic.List(ctx, &runtimeClassList, client.MatchingFields{"metadata.name": className}); err != nil {
+		return fmt.Errorf("unable to list runtime classes: %w", err)
+	}
+
+	if len(runtimeClassList.Items) < 1 {
+		spinClass := getWasmRuntimeClassTemplate(wasmSpinRuntime)
+		if err := applyRuntimeClassManifest(ctx, kube, spinClass); err != nil {
+			return fmt.Errorf("unable to apply wasm spin runtime class %q: %w", className, err)
+		}
+	}
+
+	return nil
+}
+
+func applyRuntimeClassManifest(ctx context.Context, kube *kubeclient, manifest string) error {
+	var runtimeClassObj nodev1.RuntimeClass
+	if err := yaml.Unmarshal([]byte(manifest), &runtimeClassObj); err != nil {
+		return fmt.Errorf("failed to unmarshal RuntimeClass manifest: %w", err)
+	}
+
+	desired := runtimeClassObj.DeepCopy()
+	_, err := controllerutil.CreateOrUpdate(ctx, kube.dynamic, &runtimeClassObj, func() error {
```

Comment:
had trouble getting this to work for the update scenario in the case that the runtime class already exists, kept getting errors along the lines of:

"unable to ensure wasm RuntimeClasses: unable to apply wasm spin runtime class "wasmtime-spin": failed to apply RuntimeClass manifest: runtimeclasses.node.k8s.io "wasmtime-spin" is invalid: metadata.resourceVersion: Invalid value: 0x0: must be specified for an update"

as a work around only creating if we know the runtime class doesn't already exist in the cluster

Reply:
this is sort of a hack, internally I believe it results in GET -> mutate(obj) -> PUT as update

so doing the full pointer copy basically clears all the fields that would be useful.

you can try CreateOrPatch, or you can see the "proper" usage of CreateOrUpdate over in the docs example - https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/controller/controllerutil#CreateOrPatch